Feature/login and auth

dalbanhi opened the pull request:
This will be a series of pull requests; work is still required to fully test the newly implemented features

## New Features Implemented
- [x] Implemented user, note and tag models with mongoose Schemas (see `models/`
- [ ] To do: test the creation of these
- [x] Implemented database connection (`utils/database.ts`)
- [ ] To do: test the database connection
- [x] Set up api routes for auth  using Next auth. (`/src/app/api/auth/[...nextauth]`)
- [ ] To do: test the auth with Google provider
- [x] Added shadcn to build out components for auth/log in
- [ ] To do: add a button component under `components/ui`

dalbanhi committed:
added shadcn and fixing tsconfig

dalbanhi committed:
added user and note schemas, need to add the type of tags and also test and set up mongodb

dalbanhi committed:
added tag schema

dalbanhi committed:
added database file

dalbanhi committed:
added next auth and set up Google provider with TypeScript

dalbanhi committed:
fixed comments

dalbanhi committed:
fixed merge conflict

brandonetter commented on models/user.ts:
```diff
@@ -0,0 +1,89 @@
+    minlength: 8,
+    maxlength: 30,
+    // match: /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,30}$/, // generated to capture at least 1 letter and at least one number --needs to be tested
+  },
```
Let's not store a plaintext password in our database.

We can use something like bcryptjs to handle hashing our passwords:

https://www.npmjs.com/package/bcryptjs


This way, we simply store a 1-way hashed password in our database, and when the user tries to login we can validate it by hashing the input they give us in the same way and comparing it to the hashed password in the database.

This is pretty important, as we don't want to potentially leak passwords if there's a security issue- and we definitely don't want to trust everyone who has database access with the confidential password of all of our users.

You can actually set-up the hash automatically using mongoose's 'pre' middleware.

https://mongoosejs.com/docs/middleware.html

See if you can give that a shot- otherwise we can discuss it in our meeting if we get stuck!

brandonetter commented on hooks/useMediaQuery/useDevice.ts:
You still have the error you're referencing in the stackoverflow.

Or at least I do!
![image](https://github.com/dalbanhi/git-note/assets/4108484/20cf40d7-4354-4d94-9b94-d82bcb255c61)


The rest of that answer that you're following suggests using a dynamic import for the query. I would recommend instead importing the Contribution Grid itself as a dynamic import in the `page.tsx` to solve this:

```js
// app/(main)/page.tsx
import dynamic from "next/dynamic";

const ContributionsTracker = dynamic(
  () => import("@components/shared/Home/ContributionsTracker"),
  { ssr: false }
);
import RecentPosts from "@components/shared/Home/RecentPosts";

export default function Home() {
  const userName = "Nikky Eya";
  //get only the first name
  let firstName = userName.split(" ")[0];

  return (
    <section className="flex min-h-screen w-6/12 flex-col justify-start p-4">
      <h1 className="mt-4 text-display1"> Hello, {firstName}</h1>
      <p className="mb-4 text-p1Reg text-myWhite-300">
        {"Time to jot down your latest learnings today!"}
      </p>
      <ContributionsTracker />
      <RecentPosts />
    </section>
  );
}
```

That will turn off server side rendering for this component- and you can create a loader for it if you'd like using:

```js
const ContributionsTracker = dynamic(
  () => import("@components/shared/Home/ContributionsTracker"),
  { ssr: false, loading: () => <p>some loader placeholder component here</p> }
);
```

If you're not still getting an error then we must be using some different versions of next or node. Let me know!

But if this works well for you then you don't need to worry about the custom hook, but I do like it! If we'll need all of those potential queries then let's keep the custom hook.

If you dynamically import the component we should no longer get any errors:
![image](https://github.com/dalbanhi/git-note/assets/4108484/ff4c30f1-0595-45e3-9469-c0ecac9afb55)

brandonetter commented on models/note.ts:
```diff
@@ -0,0 +1,75 @@
+  relatedNotes: [{ type: Schema.Types.ObjectId, ref: "Note" }],
+});
+
+const Note = model<INote>("Note", NoteSchema);
```
If we simply return a new model- we'll likely run into this error:

` ⨯ OverwriteModelError: Cannot overwrite `Note` model once compiled.`

The second time the model is loaded. This is because every time we try and reference this model, we'll attempt to create it again.

This is the same idea behind you using an 'isConnected' variable in database.ts

I would suggest instead importing 'models' from mongoose, so that we have access to the currently generated models- and then returning the note model if it already exists.



```suggestion
import { Schema, model, models } from "mongoose";
//....etc...
const Note = models?.Note || model<INote>("Note", NoteSchema);
```

brandonetter commented on models/tag.ts:
```diff
@@ -0,0 +1,13 @@
+  title: { type: String, required: true },
+});
+
+const Tag = model<ITag>("Tag", TagSchema);
```
```suggestion
const Tag = models?.Tag || model<ITag>("Tag", TagSchema);
```

brandonetter commented on models/user.ts:
```diff
@@ -0,0 +1,89 @@
+  socialMediaLinks: [SocialMediaLinkSchema],
+});
+
+const User = model<IUser>("User", UserSchema);
```
```suggestion
const User = models?.User || model<IUser>("User", UserSchema);
```

brandonetter commented on src/app/api/auth/[...nextauth]/route.ts:
```diff
@@ -0,0 +1,79 @@
+
+        return true;
+      } catch (e) {
+        console.log("Error signing in!!! :-( ");
```
```suggestion
        console.log("Error signing in!!! :-( ");
```
Haha

brandonetter reviewed (changes requested):
This is looking great.

A few things to get under control before we get this merged.

I'd like us to get this fully tested and working before the merge but this is a ton of progress!

Amazing work.

dalbanhi committed:
added contributions tracker dynamic import

dalbanhi committed:
added check for if the models already exist before creating them again

dalbanhi committed:
barebones auth set up with Google and Credentials

dalbanhi committed:
login forms mostly outlined

dalbanhi committed:
added continue with providers buttons

dalbanhi committed:
added reading of the user image and email and name and redirect to log-in if no session:

dalbanhi committed:
added bcrypt package for hashing pwds

dalbanhi committed:
sign in and logout working for credentials, google, and github

dalbanhi committed:
changed database connection to await/async syntax

brandonetter commented on components/shared/Form/Login/LoginForm.tsx:
```diff
@@ -0,0 +1,96 @@
+              textColor="text-myWhite-100"
+              text={`Continue with ${provider}`}
+              onClick={() => signIn(provider)}
+            ></Button>
```
This doesn't work as expected.

## 1) Signin Issues
When we call "signIn(provider)", next-auth should automatically try and connect to the given provider. Instead, we are sent to the generic login page.

This is because we're supposed to pass in "google" not "Google" and "github" not "GitHub".

```suggestion
            <Button
              key={provider}
              image={`icons/${provider}.svg`}
              backgroundColor="bg-myBlack-700"
              textColor="text-myWhite-100"
              text={`Continue with ${provider}`}
              onClick={() => signIn(provider.toLowerCase())}
            ></Button>
```

Forcing the argument to be lowercase should fix this.

## 2) Icons

This might be working for you if your computer isn't running a case-sensitive operating system. But 99% of deployment options are going to be linux, which is case-sensitive for it's file system.

That means "Google.svg" isn't going to be the same as "google.svg"

On my computer, when running in linux, the icons do not load. I would change the filenames to match what we're actually trying to load in the URL.

## 3) Uncommon Button Pattern

This isn't an error- but usually we expect to be able to use "children" in buttons, not have to send a "text" prop to them.

It would be more readable and clear to most developers if we used our button component like this:
```js
            <Button
              key={provider}
              image={`icons/${provider}.svg`}
              backgroundColor="bg-myBlack-700"
              textColor="text-myWhite-100"
              onClick={() => signIn(provider.toLowerCase())}
            >
              Continue with {provider}
            </Button>
```

Instead of with a  "text" prop. 

Give it a shot converting this to use children if you're interested.

brandonetter commented on components/shared/Form/Login/LoginForm.tsx:
```diff
@@ -0,0 +1,96 @@
+          onClick={(e) => {
+            e.preventDefault();
+            //save the result and see if there is an error, also redirect to the home page
+            signIn("credentials", {
+              email: email,
+              password: password,
+              fullName: fullName,
+            });
```
While this is fine- we can get much more control over what's happening by instead returning the result and telling it not to redirect immediately.

```js
    onClick={async (e) => {
      e.preventDefault();
      //save the result and see if there is an error, also redirect to the home page
      const result = await signIn("credentials", {
        email: email,
        password: password,
        fullName: fullName,
        redirect: false,
      });
      console.log(result);
  ```

Below, shows the 'result' on top when the password I enter is incorrect when I try and login. The bottom result is a successful login. 

![image](https://github.com/dalbanhi/git-note/assets/4108484/bb7da70d-2977-4f82-a04f-e21538406a96)

We can use these values to show an error or toast on the page letting the user know their password failed or redirect them manually.

Something to consider if we want to avoid the default "ERROR" page from next-auth 👍

brandonetter commented on components/shared/Home/ContributionsTracker.tsx:
```diff
@@ -68,8 +69,8 @@ const ContributionsTracker = () => {
-                      index * (isMedDevice ? 5 : 8) +
-                      index * (isMedDevice ? 1 : 3)
+                      index * (device?.isTablet ? 5 : 8) +
+                      index * (device?.isTablet ? 1 : 3)
```
https://www.loom.com/share/fc5456d0cff74e58854a32404a54b8c5

Let's simplify this a lot and try and just use 1 breakpoint instead of a mix of three

brandonetter reviewed (changes requested):
Key takeaways:

1) Let's double check out signIn is doing what we expect regarding the providers, and let's make sure to always have the correct filenames for everything. 

2) Let's simplify some of the changes to our ContributionsTracker to get this working consistently.

dalbanhi committed:
Bug fixes for Button, correct Sign in and prelimiary error handling on the form

dalbanhi committed:
added contributions tracker and toast from react toastify

dalbanhi committed:
added Zod validation of login form

dalbanhi commented:
Final Commits for fully implemented basic user authentication: See Loom demo here: [Loom Demo](https://www.loom.com/share/9fdbba5d2c554d1f99fee788080cedf8?sid=4e2f21ba-ee96-46bf-ac3e-c806f21698fa)

## Previous Features Implemented
- [x] Implemented user, note and tag models with mongoose Schemas (see `models/`)
- [x] Implemented database connection (`utils/database.ts`)
- [x] Set up api routes for auth  using Next auth. (`/src/app/api/auth/[...nextauth]`)

## New Features Implemented
- [x] Implemented bug fixes specified above
- [x] Implemented Login form validation with zod (see video; `lib/validators/user.schema.ts` and `components/shared/Form/Login/LoginForm.tsx`)

brandonetter commented on auth/auth.ts:
```diff
@@ -68,6 +68,13 @@ export const config = {
+          //check if the user even has a saved password
+          if (userExistsAlready.password === undefined) {
+            throw new Error(
+              "User exists but no password saved. Maybe try logging in with Google or Github?"
```
![image](https://github.com/dalbanhi/git-note/assets/4108484/92ccfa4e-d338-4bce-9df0-2e766064a793)

This is good! And you've been intentionally vague about the actual provider- which is correct here. Never give any more information than necessary.

dalbanhi committed:
removing debugging logs
